### PR TITLE
SES-4299 Use Manifold

### PR DESCRIPTION
**Pull request by efernandez**
Use `ceres::Manifold` instead of `ceres::LocalParameterization`, which was marked as deprecated in Ceres 2.1.0, and removed in 2.2.0.

This also fixes or ignores some warnings that happened with gcc 12, which includes a gcc 12 bug and a bug in a unit test.

I built all `fuse` packages successfully with these changes:
```bash
nix develop clearpath/2.32.0-20231004045951-SES-4299-0#sdk
colcon build --mixin ws-test --packages-select fuse_graphs fuse_core fuse_constraints fuse_variables fuse_loss fuse_msgs fuse_models fuse_viz fuse_publishers fuse_optimizers fuse_doc fuse_tutorials
```

All tests passed successfully:
```bash
colcon test --event-handlers console_direct+ --packages-select fuse_graphs fuse_core fuse_constraints fuse_variables fuse_loss fuse_msgs fuse_models fuse_viz fuse_publishers fuse_optimizers fuse_doc fuse_tutorials
```

:point_right: But I found a bug in the local parameterization for the 2D and 3D orientation, which were computing the jacobian for the wrong variable. The autodiff implementation was also wrong, hiding the error. The implementation in the new autodiff manifold is different. I've highlighted that with a comment in that code. This could indeed be a serious bug with broad implications.

I'll send a similar MR upstream.

### CODE REVIEW

**Comment by efernandez** [comment]
Upstream I'm going to push a bit more into simplifying `fuse_core::LocalParameterization`, to see if we can simply use `ceres::Manifold` directly instead. The only added functionatility is the Boost serialization, that I believe we don't need, because there's no state stored in the class. Indeed, I've tried to compiled without it, and everything compiles fine. This is the `diff` of the changes I made to test that out.

```diff
$ git diff
diff --git a/fuse_core/include/fuse_core/local_parameterization.h b/fuse_core/include/fuse_core/local_parameterization.h
index f8449c9..5d7ba8a 100644
--- a/fuse_core/include/fuse_core/local_parameterization.h
+++ b/fuse_core/include/fuse_core/local_parameterization.h
@@ -105,21 +105,6 @@ public:
     const double* x,
     double* jacobian) const = 0;
 #endif
-
-private:
-  // Allow Boost Serialization access to private methods
-  friend class boost::serialization::access;
-
-  /**
-   * @brief The Boost Serialize method that serializes all of the data members in to/out of the archive
-   *
-   * @param[in/out] archive - The archive object that holds the serialized class members
-   * @param[in] version - The version of the archive being read/written. Generally unused.
-   */
-  template<class Archive>
-  void serialize(Archive& /* archive */, const unsigned int /* version */)
-  {
-  }
 };
 
 }  // namespace fuse_core
diff --git a/fuse_variables/include/fuse_variables/orientation_2d_stamped.h b/fuse_variables/include/fuse_variables/orientation_2d_stamped.h
index 61269a9..4afd685 100644
--- a/fuse_variables/include/fuse_variables/orientation_2d_stamped.h
+++ b/fuse_variables/include/fuse_variables/orientation_2d_stamped.h
@@ -124,22 +124,6 @@ public:
     jacobian[0] = 1.0;
     return true;
   }
-
-private:
-  // Allow Boost Serialization access to private methods
-  friend class boost::serialization::access;
-
-  /**
-   * @brief The Boost Serialize method that serializes all of the data members in to/out of the archive
-   *
-   * @param[in/out] archive - The archive object that holds the serialized class members
-   * @param[in] version - The version of the archive being read/written. Generally unused.
-   */
-  template<class Archive>
-  void serialize(Archive& archive, const unsigned int /* version */)
-  {
-    archive & boost::serialization::base_object<fuse_core::LocalParameterization>(*this);
-  }
 };
 
 /**
diff --git a/fuse_variables/include/fuse_variables/orientation_3d_stamped.h b/fuse_variables/include/fuse_variables/orientation_3d_stamped.h
index 25db565..cfd00b2 100644
--- a/fuse_variables/include/fuse_variables/orientation_3d_stamped.h
+++ b/fuse_variables/include/fuse_variables/orientation_3d_stamped.h
@@ -156,22 +156,6 @@ public:
     jacobian[8] = -x3; jacobian[9]  =  x2; jacobian[10] = -x1;  jacobian[11] =  x0;  // NOLINT
     return true;
   }
-
-private:
-  // Allow Boost Serialization access to private methods
-  friend class boost::serialization::access;
-
-  /**
-   * @brief The Boost Serialize method that serializes all of the data members in to/out of the archive
-   *
-   * @param[in/out] archive - The archive object that holds the serialized class members
-   * @param[in] version - The version of the archive being read/written. Generally unused.
-   */
-  template<class Archive>
-  void serialize(Archive& archive, const unsigned int /* version */)
-  {
-    archive & boost::serialization::base_object<fuse_core::LocalParameterization>(*this);
-  }
 };
 
 /**
diff --git a/fuse_variables/src/orientation_2d_stamped.cpp b/fuse_variables/src/orientation_2d_stamped.cpp
index f5848c4..3f435d9 100644
--- a/fuse_variables/src/orientation_2d_stamped.cpp
+++ b/fuse_variables/src/orientation_2d_stamped.cpp
@@ -72,6 +72,5 @@ fuse_core::LocalParameterization* Orientation2DStamped::localParameterization()
 
 }  // namespace fuse_variables
 
-BOOST_CLASS_EXPORT_IMPLEMENT(fuse_variables::Orientation2DLocalParameterization);
 BOOST_CLASS_EXPORT_IMPLEMENT(fuse_variables::Orientation2DStamped);
 PLUGINLIB_EXPORT_CLASS(fuse_variables::Orientation2DStamped, fuse_core::Variable);
diff --git a/fuse_variables/src/orientation_3d_stamped.cpp b/fuse_variables/src/orientation_3d_stamped.cpp
index f5dea4a..a152bd9 100644
--- a/fuse_variables/src/orientation_3d_stamped.cpp
+++ b/fuse_variables/src/orientation_3d_stamped.cpp
@@ -76,6 +76,5 @@ fuse_core::LocalParameterization* Orientation3DStamped::localParameterization()
 
 }  // namespace fuse_variables
 
-BOOST_CLASS_EXPORT_IMPLEMENT(fuse_variables::Orientation3DLocalParameterization);
 BOOST_CLASS_EXPORT_IMPLEMENT(fuse_variables::Orientation3DStamped);
 PLUGINLIB_EXPORT_CLASS(fuse_variables::Orientation3DStamped, fuse_core::Variable);
```

The only other thing is the `FUSE_SMART_PTR_ALIASES_ONLY(LocalParameterization);` declaration, that we probably don't need either. :thinking: 

Either way, this is a note for myself, so I can start that discussion upstream. My plan is to first get this in into our fork, incorporating your review.

File: fuse_variables/include/fuse_variables/orientation_3d_stamped.h
Comment (mikepurvis) [inline fuse_variables/include/fuse_variables/orientation_3d_stamped.h:157]: This tingles my spidey senses, but I assume it makes sense to a domain expert.

Comment (efernandez) [reply fuse_variables/include/fuse_variables/orientation_3d_stamped.h:157]: Yes, this is what I mentioned in the description that looks like a latent bug hiden by two bugs cancelling each other.

I'm testing things out to confirm that with these changes things are at least fine, although I believe they should indeed be better.

The unit tests started to fail when I used the `ceres::Manifold` implementation of the code that was "duplicated" in this repo.

File: fuse_constraints/CMakeLists.txt
Comment (kriato) [inline fuse_constraints/CMakeLists.txt]: I wonder if it's better to disable the warning only for the GCC 12

Comment (efernandez) [reply fuse_constraints/CMakeLists.txt]: I can check for GCC with https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_ID.html, but I'm not sure how to check for GCC 12 specifically. Indeed, I should check for >= GCC 12, since the bug hasn't been fixed and I have no idea when it'd get fixed.

I believe for the version I'd have to use https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_VERSION.html and check for `VERSION_GREATER_EQUAL` as explained in https://cmake.org/cmake/help/latest/variable/CMAKE_VERSION.html

@kriato What do you think of the following suggestion?
```suggestion
if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 12.0)
    add_compile_options(-Wall -Werror -Wno-array-bounds)
else()
    add_compile_options(-Wall -Werror)
endif()
```

We're actually targetting gcc 12.2.0:
```bash
efernandez@CPR01878L:2.32.0-20231004045951-SES-4299-0#sdk
~$ gcc --version
gcc (GCC) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Comment (efernandez) [reply fuse_constraints/CMakeLists.txt]: Done

**Review by kriato — APPROVED** [review]
👍

**Review by mahmed-cpr — APPROVED** [review]

**Comment by efernandez** [comment]
I'm going to wait for @nlunscher-cpr to also review this MR before merging it in our fork.

I've also sent an equivalent MR upstream: https://github.com/locusrobotics/fuse/pull/341

File: fuse_core/include/fuse_core/autodiff_local_parameterization.h
Comment (efernandez) [inline fuse_core/include/fuse_core/autodiff_local_parameterization.h:215]: @nlunscher-cpr This is the main change I want you to review

File: fuse_variables/include/fuse_variables/orientation_2d_stamped.h
Comment (efernandez) [inline fuse_variables/include/fuse_variables/orientation_2d_stamped.h:125]: @nlunscher-cpr This analytic jacobian was wrong. This change goes together with fix `fuse_core::AutoDiffLocalParameterization`, that with the correct jacobian computed becomes the same as the one in `ceres::AutoDiffManifold`

File: fuse_variables/include/fuse_variables/orientation_3d_stamped.h
Comment (efernandez) [inline fuse_variables/include/fuse_variables/orientation_3d_stamped.h:138]: @nlunscher-cpr Same as above, but here it's for the 3D orientation instead of the 2D one.

The correct jacobian simply has a different sign. That's because the `Minus` operator is computing `y_minus_x = y - x`, and the jacobian was incorrectly computed wrt `y` before, and now is computed wrt `x`, which is the same jacobian but with different sign.

File: fuse_core/test/test_local_parameterization.cpp
Comment (ghost) [inline fuse_core/test/test_local_parameterization.cpp:77]:
```suggestion
    y_minus_x[0] = (y[0] - x[0]) / 2.0;
```
isn't this y - x?
Also not sure what the 2 and 5 here are doing :thinking:

**Review by ghost — APPROVED** [review]

**Comment by mikepurvis** [comment]
Can this be merged?

**Comment by efernandez** [comment]
> Can this be merged?

@mikepurvis I'm a bit afraid this could break something because the API differs from the previous one.

If there's another way to get things in, I'd like to test this out first, and also look a bit deeper into the code and how Ceres solver uses the `Minus` operator, which has changed (args are swapped).

**Comment by mikepurvis** [comment]
Okay, so it looks like all this is really doing is resolving warnings— in that case, my inclination would be to merge the nixpkgs bump without it, and then at that point it can be merged coincident with a follow-on nix-base MR that removes the `disableWerror` override for the fuse packages.

**Comment by efernandez** [comment]
Already merged upstream in https://github.com/locusrobotics/fuse/pull/341 and https://github.com/locusrobotics/fuse/pull/363

@jakemclaughlin6 also took care of syncing our fork to make those changes into our systems.
@mikepurvis The flags to disable the warnings as errors for deprecated code in fuse are not longer needed in our buildfarm.

**Pull request closed without merging**
